libevmasm/RuleList.h

```diff
@@ -349,14 +349,26 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart7(
 	rules.push_back({
 		// SHL(B, SHL(A, X)) -> SHL(min(A+B, 256), X)
 		{Instruction::SHL, {{B}, {Instruction::SHL, {{A}, {X}}}}},
-		[=]() -> Pattern { return {Instruction::SHL, {std::min(A.d() + B.d(), u256(256)), X}}; },
+		[=]() -> Pattern {
+			bigint sum = bigint(A.d()) + B.d();
+			if (sum >= 256)
+				return 0;
```

Comment:
In this case, the rule now does remove the non-constant, but it is still indicated that it doesn't by the ``false`` below. This should probably instead still return ``{Instruction::SHL, {u256(256), X}}``. Alternatively, the whole rule could rather look like:

```
[=]() -> Pattern {
  return {Instruction::SHL, {u256(std::min(bigint(A.d()) + bigint(B.d()), bigint(256))), X }};
}
```

Reply:
The non-constant can then be conditionally removed with the mechanism in #6160 independently of this rule.

Reply:
I know that what I'm suggesting was one of the previous versions of this PR and that it was then explicitly changed - but won't e.g. the yul ``ExpressionSimplifier`` remove a non-movable ``X`` in certain cases if we don't change it back? ``pop(shl(1,shl(255,call(...))))`` reducing to nothing?

Reply:
Oh dear! Good catch!

I don't understand your last message.

Reply:
If you agree that this needs to be changed back, nevermind my last comment :-).